src/webgpu/api/operation/render_pipeline/overrides.spec.ts

```diff
@@ -1,12 +1,459 @@
+        module: t.device.createShaderModule({
+          code: kFullScreenTriangleFragmentShader,
+        }),
+        entryPoint: 'main',
+        constants: t.params.fragmentConstants,
+        targets: [{ format }],
+      }
+    );
+  });
+
+g.test('shared_shader_module')
+  .desc(
+    `Test that when the same module is shared by different pipelines, the constant values are still being used correctly.`
+  )
+  .params(u =>
+    u
+      .combine('isAsync', [true, false])
+      .beginSubcases()
+      .combineWithParams([
+        {
+          expected0: kClearValueResult,
+          vertexConstants0: {
+            xright: -3.0,
+          } as Record<string, GPUPipelineConstantValue>,
+          fragmentConstants0: {},
+
+          expected1: kDefaultValueResult,
+          vertexConstants1: {},
+          fragmentConstants1: {},
+        },
+        {
+          expected0: { R: 0.0, G: 0.0, B: 0.0, A: 0.0 },
+          vertexConstants0: {},
+          fragmentConstants0: { R: 0.0, G: 0.0, B: 0.0, A: 0.0 } as Record<
+            string,
+            GPUPipelineConstantValue
+          >,
+
+          expected1: kDefaultValueResult,
+          vertexConstants1: {},
+          fragmentConstants1: {},
+        },
+        {
+          expected0: { R: 1.0, G: 0.0, B: 1.0, A: 0.0 },
+          vertexConstants0: {},
+          fragmentConstants0: { R: 1.0, G: 0.0, B: 1.0, A: 0.0 } as Record<
+            string,
+            GPUPipelineConstantValue
+          >,
+
+          expected1: { R: 0.0, G: 1.0, B: 0.0, A: 1.0 },
+          vertexConstants1: {},
+          fragmentConstants1: { R: 0.0, G: 1.0, B: 0.0, A: 1.0 } as Record<
+            string,
+            GPUPipelineConstantValue
+          >,
+        },
+      ])
+  )
+  .fn(async t => {
+    const format = 'bgra8unorm';
+    const vertexModule = t.device.createShaderModule({
+      code: kFullScreenTriangleVertexShader,
+    });
+
+    const fragmentModule = t.device.createShaderModule({
+      code: kFullScreenTriangleFragmentShader,
+    });
+
+    const createPipelineFn = async (
+      vertexConstants: Record<string, GPUPipelineConstantValue>,
+      fragmentConstants: Record<string, GPUPipelineConstantValue>
+    ) => {
+      const descriptor: GPURenderPipelineDescriptor = {
+        layout: 'auto',
+        vertex: {
+          module: vertexModule,
+          entryPoint: 'main',
+          constants: vertexConstants,
+        },
+        fragment: {
+          module: fragmentModule,
+          entryPoint: 'main',
+          targets: [{ format }],
+          constants: fragmentConstants,
+        },
+        primitive: {
+          topology: 'triangle-list',
+          frontFace: 'ccw',
+          cullMode: 'back',
+        },
+      };
+
+      const promise = t.params.isAsync
+        ? t.device.createRenderPipelineAsync(descriptor)
+        : Promise.resolve(t.device.createRenderPipeline(descriptor));
+
+      const pipeline = await promise;
+      return pipeline;
```

Comment:
µNit: Simply this should work, because of the conveniences of how `await` works:

```suggestion
      return t.params.isAsync
        ? t.device.createRenderPipelineAsync(descriptor)
        : t.device.createRenderPipeline(descriptor);
```

though the function type inference might not be happy so you can either add back the `Promise.resolve` or put an explicit return type of `GPURenderPipeline | Promise<GPURenderPipeline>` on the function